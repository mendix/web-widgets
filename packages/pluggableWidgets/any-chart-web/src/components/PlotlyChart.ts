import { PlotlyHTMLElement, Data, Layout, Config, newPlot, react, purge } from "plotly.js-dist-min";

export interface ChartProps {
    data: Data[];
    layout: Partial<Layout>;
    config: Partial<Config>;
    width: number;
    height: number;
}

export class PlotlyChart {
    private element: HTMLElement;
    private plotlyElement?: PlotlyHTMLElement;
    private data: Data[];
    private layout: Partial<Layout>;
    private config: Partial<Config>;

    constructor(element: HTMLElement, props: ChartProps) {
        this.element = element;
        this.data = props.data;
        this.layout = props.layout;
        this.config = props.config;
        this.init();
    }

    private init(): void {
        newPlot(this.element, this.data, this.layout, this.config)
            .then(plotlyElement => {
                this.plotlyElement = plotlyElement;
            })
            .catch(error => {
                console.error("Error initializing chart:", error);
            });
    }

    update(props: Partial<ChartProps>): void {
        if (props.data) {
            this.data = props.data;
        }
        if (props.layout) {
            this.layout = props.layout;
        }
        if (props.config) {
            this.config = props.config;
        }
        this.redraw();
    }

    private redraw(): void {
        if (this.plotlyElement) {
            react(this.element, this.data, this.layout, this.config).catch(error => {
                console.error("Error updating chart:", error);
            });
        }
    }

    destroy(): void {
        if (this.plotlyElement) {
            purge(this.element);
            this.plotlyElement = undefined;
        }
    }
}
