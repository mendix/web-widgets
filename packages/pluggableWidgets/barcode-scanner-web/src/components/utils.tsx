import {
    BinaryBitmap,
    BrowserMultiFormatReader,
    HTMLCanvasElementLuminanceSource,
    HybridBinarizer,
    BarcodeFormat,
    DecodeHintType,
    Result
} from "@zxing/library";
import { BarcodeFormatsType } from "typings/BarcodeScannerProps";

export const returnVideoWidthHeight = (
    curVideoRef: HTMLVideoElement,
    canvasMiddle: HTMLDivElement
): { width: number; height: number } => {
    const aspectRatioClient = curVideoRef.clientWidth / curVideoRef.clientHeight;
    const aspectRatioVideo = curVideoRef.videoWidth / curVideoRef.videoHeight;

    let videoCropWidth = (canvasMiddle.clientWidth / curVideoRef.clientWidth) * curVideoRef.videoWidth;
    let videoCropHeight = (canvasMiddle.clientHeight / curVideoRef.clientHeight) * curVideoRef.videoHeight;

    if (aspectRatioVideo < aspectRatioClient) {
        videoCropHeight = videoCropWidth;
    } else {
        videoCropWidth = videoCropHeight;
    }
    return { width: videoCropWidth, height: videoCropHeight };
};

export const drawCropOnCanvas = (
    captureCanvas: HTMLCanvasElement,
    videoRef: HTMLVideoElement,
    canvasMiddle: HTMLDivElement
): HTMLCanvasElement => {
    const canvasContext = captureCanvas.getContext("2d", { willReadFrequently: true });
    const videoWidthHeight = returnVideoWidthHeight(videoRef, canvasMiddle);
    if (
        Math.abs(captureCanvas.width - videoWidthHeight.width) > 1 ||
        Math.abs(captureCanvas.height - videoWidthHeight.height) > 1
    ) {
        captureCanvas.width = videoWidthHeight.width;
        captureCanvas.height = videoWidthHeight.height;
    }
    if (canvasContext !== null && videoRef !== null) {
        canvasContext.drawImage(
            videoRef,
            (videoRef.videoWidth - videoWidthHeight.width) / 2,
            (videoRef.videoHeight - videoWidthHeight.height) / 2,
            videoWidthHeight.width,
            videoWidthHeight.height,
            0,
            0,
            videoWidthHeight.width,
            videoWidthHeight.height
        );
    }
    return captureCanvas;
};

export function decodeBarcodeDetector(
    resolve: (value: Result | string) => void,
    reject: (reason?: Error) => void,
    captureCanvas: HTMLCanvasElement,
    videoRef: HTMLVideoElement,
    canvasMiddle: HTMLDivElement,
    barcodeDetector?: any
): void {
    const croppedOnCanvas = drawCropOnCanvas(captureCanvas, videoRef, canvasMiddle);
    barcodeDetector
        .detect(croppedOnCanvas)
        .then((barcodes: any) => {
            if (barcodes[0]?.rawValue) {
                resolve(barcodes[0]?.rawValue);
            } else {
                setTimeout(
                    () =>
                        decodeBarcodeDetector(resolve, reject, captureCanvas, videoRef, canvasMiddle, barcodeDetector),
                    50
                );
            }
        })
        .catch((err: any) => {
            reject(err);
        });
}

export const decodeCanvas = (reader: BrowserMultiFormatReader, captureCanvas: HTMLCanvasElement): Result => {
    const luminanceSource = new HTMLCanvasElementLuminanceSource(captureCanvas);

    const binaryBitmap = new BinaryBitmap(new HybridBinarizer(luminanceSource));
    const result = reader.decodeBitmap(binaryBitmap);
    return result;
};

export const createHints = (
    useAllFormats: boolean,
    barcodeFormats?: BarcodeFormatsType[]
): Map<DecodeHintType, any> => {
    const hints = new Map();
    let formats: BarcodeFormat[] = [];
    if (useAllFormats) {
        formats = [
            BarcodeFormat.UPC_A,
            BarcodeFormat.UPC_E,
            BarcodeFormat.EAN_8,
            BarcodeFormat.EAN_13,
            BarcodeFormat.CODE_39,
            BarcodeFormat.CODE_128,
            BarcodeFormat.ITF,
            BarcodeFormat.RSS_14,
            BarcodeFormat.QR_CODE,
            BarcodeFormat.DATA_MATRIX,
            BarcodeFormat.AZTEC,
            BarcodeFormat.PDF_417
        ];
    } else {
        if (barcodeFormats) {
            formats = barcodeFormats.map(val => BarcodeFormat[val.barcodeFormat]);
        }
    }
    hints.set(DecodeHintType.POSSIBLE_FORMATS, formats);
    hints.set(DecodeHintType.ENABLE_CODE_39_EXTENDED_MODE, true);
    return hints;
};
