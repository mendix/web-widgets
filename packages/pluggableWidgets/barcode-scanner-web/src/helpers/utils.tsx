import {
    BarcodeFormat,
    BinaryBitmap,
    BrowserMultiFormatReader,
    DecodeHintType,
    HTMLCanvasElementLuminanceSource,
    HybridBinarizer,
    Result
} from "@zxing/library";
import { RefObject } from "react";
import { BarcodeFormatsType } from "typings/BarcodeScannerProps";

export const mediaStreamConstraints: MediaStreamConstraints = {
    audio: false,
    video: {
        facingMode: "environment",
        width: { min: 1280, ideal: 4096, max: 4096 },
        height: { min: 720, ideal: 2160, max: 2160 }
    }
};

export type ReaderProps = {
    onSuccess?: (data: string) => void;
    onError?: (e: Error) => void;
    useCrop: boolean;
    barcodeFormats?: BarcodeFormatsType[];
    useAllFormats: boolean;
    canvasMiddleRef: RefObject<HTMLDivElement>;
    detectionLogic?: "zxing" | "native";
};

export type UseReaderHook = (args: ReaderProps) => { ref: RefObject<HTMLVideoElement>; useBrowserAPI: boolean };

export const returnVideoWidthHeight = (
    curVideoRef: HTMLVideoElement,
    canvasMiddle: HTMLDivElement
): { width: number; height: number } => {
    const aspectRatioClient = curVideoRef.clientWidth / curVideoRef.clientHeight;
    const aspectRatioVideo = curVideoRef.videoWidth / curVideoRef.videoHeight;

    let videoCropWidth = (canvasMiddle.clientWidth / curVideoRef.clientWidth) * curVideoRef.videoWidth;
    let videoCropHeight = (canvasMiddle.clientHeight / curVideoRef.clientHeight) * curVideoRef.videoHeight;

    if (aspectRatioVideo < aspectRatioClient) {
        videoCropHeight = videoCropWidth;
    } else {
        videoCropWidth = videoCropHeight;
    }
    return { width: videoCropWidth, height: videoCropHeight };
};

export const drawCropOnCanvas = (
    captureCanvas: HTMLCanvasElement,
    videoRef: HTMLVideoElement,
    canvasMiddle: HTMLDivElement
): HTMLCanvasElement => {
    const canvasContext = captureCanvas.getContext("2d", { willReadFrequently: true });
    const videoWidthHeight = returnVideoWidthHeight(videoRef, canvasMiddle);
    if (
        Math.abs(captureCanvas.width - videoWidthHeight.width) > 1 ||
        Math.abs(captureCanvas.height - videoWidthHeight.height) > 1
    ) {
        captureCanvas.width = videoWidthHeight.width;
        captureCanvas.height = videoWidthHeight.height;
    }
    if (canvasContext !== null && videoRef !== null) {
        canvasContext.drawImage(
            videoRef,
            (videoRef.videoWidth - videoWidthHeight.width) / 2,
            (videoRef.videoHeight - videoWidthHeight.height) / 2,
            videoWidthHeight.width,
            videoWidthHeight.height,
            0,
            0,
            videoWidthHeight.width,
            videoWidthHeight.height
        );
    }
    return captureCanvas;
};

export const decodeCanvas = (reader: BrowserMultiFormatReader, captureCanvas: HTMLCanvasElement): Result => {
    const luminanceSource = new HTMLCanvasElementLuminanceSource(captureCanvas);

    const binaryBitmap = new BinaryBitmap(new HybridBinarizer(luminanceSource));
    const result = reader.decodeBitmap(binaryBitmap);
    return result;
};

export const createHints = (
    useAllFormats: boolean,
    barcodeFormats?: BarcodeFormatsType[]
): Map<DecodeHintType, any> => {
    const hints = new Map();
    let formats: BarcodeFormat[] = [];
    if (useAllFormats) {
        formats = [
            BarcodeFormat.UPC_A,
            BarcodeFormat.UPC_E,
            BarcodeFormat.EAN_8,
            BarcodeFormat.EAN_13,
            BarcodeFormat.CODE_39,
            BarcodeFormat.CODE_128,
            BarcodeFormat.ITF,
            BarcodeFormat.RSS_14,
            BarcodeFormat.QR_CODE,
            BarcodeFormat.DATA_MATRIX,
            BarcodeFormat.AZTEC,
            BarcodeFormat.PDF_417
        ];
    } else {
        if (barcodeFormats) {
            formats = barcodeFormats.map(val => BarcodeFormat[val.barcodeFormat]);
        }
    }
    hints.set(DecodeHintType.POSSIBLE_FORMATS, formats);
    hints.set(DecodeHintType.ENABLE_CODE_39_EXTENDED_MODE, true);
    return hints;
};
