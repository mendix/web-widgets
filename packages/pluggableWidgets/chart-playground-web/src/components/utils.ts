function isNonEmptyString(value: undefined | null | string): value is string {
    return value !== null && value !== undefined && value !== "";
}

export const emptyObjectString = "{}";

export function ifNonEmptyStringElseEmptyObjectString(value: undefined | null | string): string {
    return isNonEmptyString(value) ? value : emptyObjectString;
}
