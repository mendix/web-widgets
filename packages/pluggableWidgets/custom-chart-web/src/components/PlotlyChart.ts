import { PlotlyHTMLElement, Data, Layout, Config, newPlot, react, purge } from "plotly.js-dist-min";

export interface ChartProps {
    data: Data[];
    layout: Partial<Layout>;
    config: Partial<Config>;
    width: number;
    height: number;
    onClick?: (data: any) => void;
}

export class PlotlyChart {
    private element: HTMLElement;
    private plotlyElement?: PlotlyHTMLElement;
    private data: Data[];
    private layout: Partial<Layout>;
    private config: Partial<Config>;

    constructor(element: HTMLElement, props: ChartProps) {
        this.element = element;
        this.data = props.data;
        this.layout = props.layout;
        this.config = props.config;
        this.init(props);
    }

    private init(props: ChartProps): void {
        newPlot(this.element, this.data, this.layout, this.config)
            .then(plotlyElement => {
                this.plotlyElement = plotlyElement;
                if (props?.onClick) {
                    this.plotlyElement.on("plotly_click", props.onClick);
                }
            })
            .catch(error => {
                console.error("Error initializing chart:", error);
            });
    }

    update(props: Partial<ChartProps>): void {
        if (props.data) {
            this.data = props.data;
        }
        if (props.layout) {
            this.layout = props.layout;
        }
        if (props.config) {
            this.config = props.config;
        }
        this.redraw();
    }

    private redraw(): void {
        if (this.plotlyElement) {
            react(this.element, this.data, this.layout, this.config).catch(error => {
                console.error("Error updating chart:", error);
            });
        }
    }

    destroy(): void {
        if (this.plotlyElement) {
            purge(this.element);
            this.plotlyElement = undefined;
        }
    }
}
