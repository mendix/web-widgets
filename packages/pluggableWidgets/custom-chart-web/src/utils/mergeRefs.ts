import { MutableRefObject, Ref, RefCallback } from "react";

export function mergeRefs<T>(...refs: Array<Ref<T>>): Ref<T> | RefCallback<T> | undefined {
    if (refs.length === 0) {
        return undefined;
    } else if (refs.length === 1) {
        return refs[0];
    } else {
        return ref => {
            for (const x of refs) {
                if (typeof x === "function") {
                    x(ref);
                } else if (x) {
                    (x as MutableRefObject<T | null>).current = ref;
                }
            }
        };
    }
}
