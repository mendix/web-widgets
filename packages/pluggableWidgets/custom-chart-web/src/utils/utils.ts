import { EditorStoreState } from "@mendix/shared-charts/main";
import { Config, Data, Layout } from "plotly.js-dist-min";
import { ChartProps } from "../components/PlotlyChart";

export function parseData(staticData?: string, attributeData?: string, sampleData?: string): Data[] {
    let finalData: Data[] = [];

    try {
        const dataAttribute = attributeData ? JSON.parse(attributeData) : [];
        finalData = [...finalData, ...(staticData ? JSON.parse(staticData) : []), ...dataAttribute];

        if (dataAttribute.length === 0) {
            finalData = [...finalData, ...(sampleData ? JSON.parse(sampleData) : [])];
        }
    } catch (error) {
        console.error("Error parsing chart data:", error);
    }

    return finalData;
}

export function parseLayout(staticLayout?: string, attributeLayout?: string, sampleLayout?: string): Partial<Layout> {
    let finalLayout: Partial<Layout> = {};

    try {
        const layoutAttribute = attributeLayout ? JSON.parse(attributeLayout) : {};
        finalLayout = { ...finalLayout, ...(staticLayout ? JSON.parse(staticLayout) : {}), ...layoutAttribute };

        if (Object.keys(layoutAttribute).length === 0) {
            finalLayout = { ...finalLayout, ...(sampleLayout ? JSON.parse(sampleLayout) : {}) };
        }
    } catch (error) {
        console.error("Error parsing chart layout:", error);
    }
    return finalLayout;
}

export function parseConfig(configOptions?: string): Partial<Config> {
    if (!configOptions) {
        return {};
    }

    try {
        return JSON.parse(configOptions);
    } catch (error) {
        console.error("Error parsing chart config:", error);
        return {};
    }
}

export function mergeChartProps(chartProps: ChartProps, editorState: EditorStoreState): ChartProps {
    return {
        ...chartProps,
        config: {
            ...chartProps.config,
            ...parseConfig(editorState.config)
        },
        layout: {
            ...chartProps.layout,
            ...parseLayout(editorState.layout)
        },
        data: chartProps.data.map((trace, index) => {
            let stateTrace: Data = {};
            try {
                if (!editorState.data || !editorState.data[index]) {
                    return trace;
                }
                stateTrace = JSON.parse(editorState.data[index]);
            } catch {
                console.warn(`Editor props for trace(${index}) is not a valid JSON:${editorState.data[index]}`);
                console.warn("Please make sure the props is a valid JSON string.");
            }
            return {
                ...trace,
                ...stateTrace
            } as Data;
        })
    };
}
