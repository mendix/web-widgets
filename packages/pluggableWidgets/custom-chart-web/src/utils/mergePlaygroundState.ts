import { EditorStoreState } from "@mendix/shared-charts/main";
import { ChartProps } from "../components/PlotlyChart";
import { parseConfig, parseLayout } from "./utils";

export function mergePlaygroundState(props: ChartProps, state: EditorStoreState): ChartProps {
    return {
        ...props,
        config: {
            ...props.config,
            ...parseConfig(state.config)
        },
        layout: {
            ...props.layout,
            ...parseLayout(state.layout)
        },
        data: props.data.map((trace, index) => ({ ...trace, customSeriesOptions: state.data[index] }))
    };
}
