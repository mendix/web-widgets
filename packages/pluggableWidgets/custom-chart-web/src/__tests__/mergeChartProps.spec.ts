import { EditorStoreState } from "@mendix/shared-charts/main";
import { ChartProps } from "../components/PlotlyChart";
import { mergeChartProps } from "../utils/utils";

// Mock console.warn to capture warnings
let consoleWarnSpy: jest.SpyInstance;
let consoleMockCalls: string[][];

beforeEach(() => {
    consoleMockCalls = [];
    consoleWarnSpy = jest.spyOn(console, "warn").mockImplementation((...args: any[]) => {
        consoleMockCalls.push(args.map(arg => String(arg)));
    });
});

afterEach(() => {
    consoleWarnSpy.mockRestore();
});

describe("mergeChartProps utility function", () => {
    describe("JSON parsing failures generate console warnings", () => {
        it("should warn for invalid JSON in editor state data", () => {
            // Arrange
            const chartProps: ChartProps = {
                data: [{ x: [1, 2], y: [3, 4], type: "scatter" }],
                layout: { title: "Test Chart" },
                config: { displayModeBar: false },
                width: 800,
                height: 600
            };

            const editorState: EditorStoreState = {
                layout: "{}",
                config: "{}",
                data: ["invalid json"]
            };

            // Act
            mergeChartProps(chartProps, editorState);

            // Assert
            expect(consoleMockCalls).toMatchSnapshot();
        });

        it("should warn for multiple invalid JSON entries", () => {
            // Arrange
            const chartProps: ChartProps = {
                data: [
                    { x: [1], y: [2], type: "scatter" },
                    { x: [3], y: [4], type: "bar" }
                ],
                layout: {},
                config: {},
                width: 800,
                height: 600
            };

            const editorState: EditorStoreState = {
                layout: "{}",
                config: "{}",
                data: ["invalid json", "{broken: json}"]
            };

            // Act
            mergeChartProps(chartProps, editorState);

            // Assert
            expect(consoleMockCalls).toMatchSnapshot();
        });

        it("should not warn when editor state has undefined entries", () => {
            // Arrange
            const chartProps: ChartProps = {
                data: [{ x: [1], y: [2], type: "scatter" }],
                layout: {},
                config: {},
                width: 800,
                height: 600
            };

            const editorState: EditorStoreState = {
                layout: "{}",
                config: "{}",
                data: [] // Empty array, so data[0] will be undefined
            };

            // Act
            mergeChartProps(chartProps, editorState);

            // Assert
            expect(consoleMockCalls).toMatchSnapshot();
        });

        it("should not warn for valid JSON", () => {
            // Arrange
            const chartProps: ChartProps = {
                data: [{ x: [1], y: [2], type: "scatter" }],
                layout: {},
                config: {},
                width: 800,
                height: 600
            };

            const editorState: EditorStoreState = {
                layout: "{}",
                config: "{}",
                data: ['{"marker": {"color": "red"}}']
            };

            // Act
            mergeChartProps(chartProps, editorState);

            // Assert
            expect(consoleMockCalls).toMatchSnapshot();
        });

        it("should handle mixed valid and invalid JSON", () => {
            // Arrange
            const chartProps: ChartProps = {
                data: [
                    { x: [1], y: [2], type: "scatter" },
                    { x: [3], y: [4], type: "bar" },
                    { x: [5], y: [6], type: "scatter" as const }
                ],
                layout: {},
                config: {},
                width: 800,
                height: 600
            };

            const editorState: EditorStoreState = {
                layout: "{}",
                config: "{}",
                data: [
                    '{"marker": {"color": "red"}}', // Valid
                    "invalid json", // Invalid
                    '{"marker": {"color": "blue"}}' // Valid
                ]
            };

            // Act
            mergeChartProps(chartProps, editorState);

            // Assert
            expect(consoleMockCalls).toMatchSnapshot();
        });
    });

    describe("Data merging behavior", () => {
        it("should merge valid editor state data with chart props", () => {
            // Arrange
            const chartProps: ChartProps = {
                data: [{ x: [1, 2], y: [3, 4], type: "scatter", name: "original" }],
                layout: { title: "Original Title" },
                config: { displayModeBar: true },
                width: 800,
                height: 600
            };

            const editorState: EditorStoreState = {
                layout: '{"title": "Modified Title"}',
                config: '{"displayModeBar": false}',
                data: ['{"marker": {"color": "red"}, "name": "modified"}']
            };

            // Act
            const result = mergeChartProps(chartProps, editorState);

            // Assert
            expect(result.data[0]).toEqual({
                x: [1, 2],
                y: [3, 4],
                type: "scatter",
                name: "modified", // Overridden by editor state
                marker: { color: "red" } // Added by editor state
            });
            expect(result.layout.title).toBe("Modified Title");
            expect(result.config.displayModeBar).toBe(false);
        });

        it("should use empty object for invalid JSON entries", () => {
            // Arrange
            const chartProps: ChartProps = {
                data: [{ x: [1], y: [2], type: "scatter", name: "original" }],
                layout: {},
                config: {},
                width: 800,
                height: 600
            };

            const editorState: EditorStoreState = {
                layout: "{}",
                config: "{}",
                data: ["invalid json"]
            };

            // Act
            const result = mergeChartProps(chartProps, editorState);

            // Assert
            expect(result.data[0]).toEqual({
                x: [1],
                y: [2],
                type: "scatter",
                name: "original" // Original properties preserved
            });
        });

        it("should only process traces that exist in chart props", () => {
            // Arrange
            const chartProps: ChartProps = {
                data: [{ x: [1], y: [2], type: "scatter" }], // Only 1 trace
                layout: {},
                config: {},
                width: 800,
                height: 600
            };

            const editorState: EditorStoreState = {
                layout: "{}",
                config: "{}",
                data: [
                    '{"marker": {"color": "red"}}', // Will be processed
                    '{"marker": {"color": "blue"}}', // Will be ignored
                    "invalid json" // Will be ignored
                ]
            };

            // Act
            const result = mergeChartProps(chartProps, editorState);

            // Assert
            expect(result.data).toHaveLength(1); // Only 1 trace in result
            expect(result.data[0]).toEqual({
                x: [1],
                y: [2],
                type: "scatter",
                marker: { color: "red" }
            });
            expect(consoleMockCalls).toMatchSnapshot(); // No warnings for ignored entries
        });

        it("should preserve original trace data when editor state has missing or falsy entries", () => {
            // Arrange
            const chartProps: ChartProps = {
                data: [
                    { x: [1], y: [1], type: "scatter", name: "original1", marker: { color: "blue" } },
                    { x: [2], y: [2], type: "bar", name: "original2" },
                    { x: [3], y: [3], type: "scatter", name: "original3" }
                ],
                layout: {},
                config: {},
                width: 800,
                height: 600
            };

            const editorState: EditorStoreState = {
                layout: "{}",
                config: "{}",
                data: [""] // Only one entry (empty string), so traces 1 and 2 will get undefined
            };

            // Act
            const result = mergeChartProps(chartProps, editorState);

            // Assert - All original trace data should be preserved
            expect(result.data[0]).toEqual({
                x: [1],
                y: [1],
                type: "scatter",
                name: "original1",
                marker: { color: "blue" }
            });
            expect(result.data[1]).toEqual({ x: [2], y: [2], type: "bar", name: "original2" });
            expect(result.data[2]).toEqual({ x: [3], y: [3], type: "scatter", name: "original3" });
            expect(consoleMockCalls).toMatchSnapshot();
        });
    });

    describe("Edge cases", () => {
        it("should handle empty chart props data", () => {
            // Arrange
            const chartProps: ChartProps = {
                data: [], // No traces
                layout: {},
                config: {},
                width: 800,
                height: 600
            };

            const editorState: EditorStoreState = {
                layout: "{}",
                config: "{}",
                data: ["invalid json"] // Will be ignored
            };

            // Act
            const result = mergeChartProps(chartProps, editorState);

            // Assert
            expect(result.data).toHaveLength(0);
            expect(consoleMockCalls).toMatchSnapshot();
        });

        it("should handle null JSON parsing", () => {
            // Arrange
            const chartProps: ChartProps = {
                data: [{ x: [1], y: [2], type: "scatter" }],
                layout: {},
                config: {},
                width: 800,
                height: 600
            };

            const editorState: EditorStoreState = {
                layout: "{}",
                config: "{}",
                data: ["null"] // Valid JSON that parses to null
            };

            // Act
            const result = mergeChartProps(chartProps, editorState);

            // Assert
            expect(result.data[0]).toEqual({
                x: [1],
                y: [2],
                type: "scatter"
                // null gets merged, but doesn't add properties
            });
            expect(consoleMockCalls).toMatchSnapshot();
        });

        it("should handle various malformed JSON types", () => {
            // Arrange
            const chartProps: ChartProps = {
                data: [
                    { x: [1], y: [1], type: "scatter" },
                    { x: [2], y: [2], type: "bar" },
                    { x: [3], y: [3], type: "scatter" as const },
                    { x: [4], y: [4], type: "scatter" }
                ],
                layout: {},
                config: {},
                width: 800,
                height: 600
            };

            const editorState: EditorStoreState = {
                layout: "{}",
                config: "{}",
                data: [
                    '{"valid": "json"}', // Valid
                    "{invalid json}", // Invalid
                    "", // Invalid empty string - treated as falsy, early return
                    "undefined" // Invalid
                ]
            };

            // Act
            mergeChartProps(chartProps, editorState);

            // Assert
            expect(consoleMockCalls).toMatchSnapshot();
        });

        it("should handle falsy values in editor state data without warnings", () => {
            // Arrange
            const chartProps: ChartProps = {
                data: [
                    { x: [1], y: [1], type: "scatter", name: "trace1" },
                    { x: [2], y: [2], type: "bar", name: "trace2" },
                    { x: [3], y: [3], type: "scatter", name: "trace3" }
                ],
                layout: {},
                config: {},
                width: 800,
                height: 600
            };

            const editorState: EditorStoreState = {
                layout: "{}",
                config: "{}",
                data: [
                    "", // Empty string - falsy
                    null as any, // null - falsy
                    undefined as any // undefined - falsy
                ]
            };

            // Act
            const result = mergeChartProps(chartProps, editorState);

            // Assert - No warnings should be generated for falsy values
            expect(consoleMockCalls).toMatchSnapshot();
            // Original traces should be returned unchanged
            expect(result.data[0]).toEqual({ x: [1], y: [1], type: "scatter", name: "trace1" });
            expect(result.data[1]).toEqual({ x: [2], y: [2], type: "bar", name: "trace2" });
            expect(result.data[2]).toEqual({ x: [3], y: [3], type: "scatter", name: "trace3" });
        });
    });
});
