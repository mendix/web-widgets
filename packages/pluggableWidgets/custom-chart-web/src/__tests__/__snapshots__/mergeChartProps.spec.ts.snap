// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mergeChartProps utility function Data merging behavior should only process traces that exist in chart props 1`] = `[]`;

exports[`mergeChartProps utility function Data merging behavior should preserve original trace data when editor state has missing or falsy entries 1`] = `[]`;

exports[`mergeChartProps utility function Edge cases should handle empty chart props data 1`] = `[]`;

exports[`mergeChartProps utility function Edge cases should handle falsy values in editor state data without warnings 1`] = `[]`;

exports[`mergeChartProps utility function Edge cases should handle null JSON parsing 1`] = `[]`;

exports[`mergeChartProps utility function Edge cases should handle various malformed JSON types 1`] = `
[
  [
    "Editor props for trace(1) is not a valid JSON:{invalid json}",
  ],
  [
    "Please make sure the props is a valid JSON string.",
  ],
  [
    "Editor props for trace(3) is not a valid JSON:undefined",
  ],
  [
    "Please make sure the props is a valid JSON string.",
  ],
]
`;

exports[`mergeChartProps utility function JSON parsing failures generate console warnings should handle mixed valid and invalid JSON 1`] = `
[
  [
    "Editor props for trace(1) is not a valid JSON:invalid json",
  ],
  [
    "Please make sure the props is a valid JSON string.",
  ],
]
`;

exports[`mergeChartProps utility function JSON parsing failures generate console warnings should not warn for valid JSON 1`] = `[]`;

exports[`mergeChartProps utility function JSON parsing failures generate console warnings should not warn when editor state has undefined entries 1`] = `[]`;

exports[`mergeChartProps utility function JSON parsing failures generate console warnings should warn for invalid JSON in editor state data 1`] = `
[
  [
    "Editor props for trace(0) is not a valid JSON:invalid json",
  ],
  [
    "Please make sure the props is a valid JSON string.",
  ],
]
`;

exports[`mergeChartProps utility function JSON parsing failures generate console warnings should warn for multiple invalid JSON entries 1`] = `
[
  [
    "Editor props for trace(0) is not a valid JSON:invalid json",
  ],
  [
    "Please make sure the props is a valid JSON string.",
  ],
  [
    "Editor props for trace(1) is not a valid JSON:{broken: json}",
  ],
  [
    "Please make sure the props is a valid JSON string.",
  ],
]
`;
